rfcs/20200306-single-client-parameter-server.md

```diff
@@ -0,0 +1,517 @@
+*   call `strategy.join` to wait until all scheduled functions are executed.
+
+```python
+# Connect to remote servers with a user-provided `ClusterResolver` object.
+strategy = ParameterServerStrategyV2(cluster_resolver)
+
+dataset_fn = # a function that returns a dataset
+
+# Clone the dataset on all workers, shuffled with different seeds.
+distributed_dataset = strategy.experimental_distribute_datasets_from_function(
+      dataset_fn)
+
+with strategy.scope():
+  # Create variables on parameter servers in a round-robin fashion.
+  model = create_model()
+  optimizer = tf.keras.optimizers.Adam()
+  accuracy = tf.keras.metrics.CategoricalAccuracy(name="train_accuracy")
+
+  @tf.function
+  def replica_fn(iterator):
+    x, y = next(iterator)
+    with tf.GradientTape() as tape:
+      predictions = model(x, table, training=True)
+      loss = compute_loss(y, predictions)
+    gradients = tape.gradient(loss, model.trainable_variables)
+    optimizer.apply_gradients(zip(gradients, model.trainable_variables))
+    accuracy.update_state(y, predictions)
+    return loss
+
+  for _ in range(num_epoches):
+    distributed_iter = iter(distributed_dataset)
+    for i in range(steps_per_epoch):
+      # strategy.schedule pushes a closure in the scheduling queue and
+      # returns a list of future objects immediately.
+      loss = strategy.schedule(replica_fn,
+                               args=(distributed_iter,))
+    strategy.join()
+    model.save()  # save checkpoint/summary...
+    print ("Loss = %f, accuracy = %f" % (
+        strategy.local_results(loss), accuracy.result()))
+```
+
+
+##### Alternative training loop: fully async
+
+Another option from calling `join` every epoch, users can choose to schedule all steps and then asynchronously print metric values. This option doesn’t require any synchronization in epoch boundaries.
+
+
+```python
+# … omitted
+with strategy.scope():
+  # … omitted
+  for _ in range(total_steps)):
+    strategy.schedule(step_fn, args=(iterators,))
+
+  # Print accuracy value every one minute.
+  while not strategy.done():
+    print("Current accuracy: %f" % accuracy.result())
+    time.sleep(60)
+# … omitted
+```
+
+#### Error Reporting From `replica_fn`
+
+Because of the non-blocking `schedule`, any exception raised in `replica_fn` wouldn’t be returned to users immediately. Actually an exception may pollute arbitrary number of functions in flight following the culprit function. We will set the error in returned `Future` objects for the culprit function and these polluted functions and we will raise exceptions when `join` is called.
+
+Therefore the best practice for users is to avoid writing any code that may raise in `replica_fn`:
+
+*   use repeated dataset so `OutOfRangeError` will be avoided;
+*   avoid using assertion ops or some debugging ops like `tf.debugging.check_numerics`.
```

Comment:
And instead do what?

Reply:
I think the best solution is being able to transport error message/stack trace from worker threads to main thread (or an error handling thread) to report to user and stop the program.

Reply:
Users can do the assertion themselves in the function and return something indicating a failure, instead of throwing.